security/s2a/internal/crypter/counter_test.go

```diff
@@ -0,0 +1,105 @@
+package crypter
+
+import (
+	"math"
+	"testing"
+)
+
+// counterFromValue creates a new counter given an initial value.
+func counterFromValue(value uint64) counter {
+	c := newCounter()
+	c.value = value
+	return c
+}
+
+func TestCounterInc(t *testing.T) {
+	for _, tc := range []struct {
+		desc                     string
+		counter, expectedCounter uint64
+		overflow                 bool
+	}{
+		{
+			desc:            "basic 1",
+			counter:         0,
+			expectedCounter: 1,
+		},
+		{
+			desc:            "basic 2",
+			counter:         123,
+			expectedCounter: 124,
+		},
+		{
+			desc:            "almost overflow",
+			counter:         math.MaxUint64 - 1,
+			expectedCounter: math.MaxUint64,
+		},
+		{
+			desc:     "max overflow",
+			counter:  math.MaxUint64,
+			overflow: true,
+		},
+	} {
+		t.Run(tc.desc, func(t *testing.T) {
+			c := counterFromValue(tc.counter)
+			c.inc()
+			val, err := c.val()
+			if tc.overflow {
```

Comment:
How about change the code to look like the following:

```
if got, want := (err == nil), !tc.overflow; got != want {
    t..Errorf("counter starting with %v, val()=(err=nil)=%v, want %v", tc.counter, got, want)
}
if err != nil {
    // compare the returned with tc.expectedCounter.
}
```

Also, you might want to consider using the following format in comparisons:

```
if got, want := gotValue, wantValue; got != want {
    t.Errorf("%v: func(%v, %v, ...) = %v, want %v", desc, param1, param2, ..., got, want)
}
```

You're already doing most of the above. I'm only referring to defining `got` and `want`. The benefit is in the error message it's straightforward what's `got` and what's `want`. We can do any necessary cleanup in other files in a separate PR.

Reply:
Done. I omitted `desc` for your suggestion since that is already given to the Go subtest framework and should print out the description already.

Reply:
That's totally fine. The error message format I mentioned is the complete one. It can be adjusted as needed. Thanks for the change.